### PR TITLE
[CALCITE-6309] Add REGEXP_LIKE function (enabled in MySQL, Oracle, PostgreSQL and Spark libraries)

### PR DESCRIPTION
**Pull request by jduo**
PostgreSQL supports a 3-argument REGEXP_LIKE variant that takes in a third optional string of regex flags.

### CODE REVIEW

**Comment by jduo** [comment]
@tanclary @julianhyde , initial PR to demonstrate the workflow. What's left:
- Need to add new tests for the 3-argument variant of REGEXP_LIKE.
- Figure out how to deal with flags that are in PG, but not in Java's Pattern class.
- Resolve differences between PG's description of the 'm' flag and Pattern's 'MULTILINE' flag. Determine if the makeFlags() function used in regex_replace() can be used here.

**Comment by mihaibudiu** [comment]
The results are checked when the tests are invoked from CalciteSqlOperatorFixture.

**Comment by jduo** [comment]
Oh, that explains it. I was running tests in SqlOperatorTest from the IntelliJ UI and was seeing that no validation was happening. I'll update the PR to remove use of Fixtures.forOperators()

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 1]
```diff
@@ -680,8 +687,15 @@ private static int makeRegexpFlags(String stringFlags) {
           flags |= Pattern.DOTALL;
           break;
         case 'm':
+          // PostgreSQL should actually interpret m to be a synonym for n, but this is
+          // relaxed for consistency.
           flags |= Pattern.MULTILINE;
           break;
+        case 's':
```
[the remark this answers is not in the document]
Comment (jduo) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:694]: It was decided (in the comments for the ticket) that we'll start with a base set of flags. There are many that can't be implemented through java.util.regex.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -400,11 +400,11 @@ Pattern toPattern() {
 
     /** Validate regex arguments in REGEXP_* fns, throws an exception
      * for invalid regex patterns, else returns a Pattern object. */
-    private Pattern validateRegexPattern(String regex, String methodName) {
+    private Pattern validateRegexPattern(String regex, String methodName, int flags) {
```
[the remark this answers is not in the document]
Comment (jduo) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:403]: Sounds good. I'll make this change. Thanks for the review @YiwenWu .

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 3]
```diff
@@ -557,10 +557,10 @@ static RelDataType deriveTypeSplit(SqlOperatorBinding operatorBinding,
           SqlFunctionCategory.STRING);
 
   /** The "REGEXP_LIKE(value, regexp)" function, equivalent to {@link #RLIKE}. */
-  @LibraryOperator(libraries = {SPARK})
+  @LibraryOperator(libraries = {SPARK, MYSQL, POSTGRESQL, ORACLE})
```
[the remark this answers is not in the document]
Comment (YiwenWu) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:560]: Understood, thank you for the clarification

File: site/_docs/reference.md [hunk 4]
```diff
@@ -2820,7 +2820,7 @@ In the following:
...
 | b | REGEXP_EXTRACT_ALL(string, regexp)             | Returns an array of all substrings in *string* that matches the *regexp*. Returns an empty array if there is no match
 | b | REGEXP_INSTR(string, regexp [, position [, occurrence [, occurrence_position]]]) | Returns the lowest 1-based position of the substring in *string* that matches the *regexp*, starting search at *position* (default 1), and until locating the nth *occurrence* (default 1). Setting occurrence_position (default 0) to 1 returns the end position of substring + 1. Returns 0 if there is no match
-| s | REGEXP_LIKE(string, regexp)                    | Equivalent to `string1 RLIKE string2`
+| m o p s | REGEXP_LIKE(string, regexp [, flags])        | Equivalent to `string1 RLIKE string2` with an optional parameter for search flags. Supported flags are: <ul><li>i: case-insensitive matching</li><li>c: case-sensitive matching</li><li>n: newline-sensitive matching</li><li>s: non-newline-sensitive matching</li><li>m: multi-line</li></ul>
```
[the remark this answers is not in the document]
Comment (YiwenWu) [reply site/_docs/reference.md:2824]: Maybe there is a better Function description here, without emphasizing `Equivalent to string1 RLIKE string2`.  
But this is a small suggestion that can be ignored.

**Review by YiwenWu — APPROVED** [review]

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 5]
```diff
@@ -3759,6 +3764,31 @@ static void checkRlikeFails(SqlOperatorFixture f) {
...
+      f1.checkBoolean("REGEXP_LIKE('ateststr', 'TEST', 'c')", false);
+      f1.checkBoolean("REGEXP_LIKE('atest\nstr', 'test.str', '')", false);
+      f1.checkBoolean("REGEXP_LIKE('atest\nstr', 'test.str', 'n')", true);
+      f1.checkBoolean("REGEXP_LIKE('atest\nstr', 'TEST.str', 'in')", true);
+      f1.checkBoolean("REGEXP_LIKE('ateststring', 'teststr', '')", true);
+      f1.checkBoolean("REGEXP_LIKE('ateststring', 'TESTstr', 'ic')", false);
+      f1.checkBoolean("REGEXP_LIKE('ateststring', 'TESTstr', 'ci')", true);
+      f1.checkBoolean("REGEXP_LIKE('atest\nstr', 'test.str', 's')", false);
+      f1.checkBoolean("REGEXP_LIKE('atest\nstr', 'test.str', 'ns')", false);
+      f1.checkBoolean("REGEXP_LIKE('atest\nstr', 'test.str', 'sn')", true);
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:3785]: Please add checknullTest

Comment (jduo) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:3785]: Done.

**Comment by mihaibudiu** [comment]
I see that this PR is approved, is it ready to merge?

**Comment by jduo** [comment]
> I see that this PR is approved, is it ready to merge?

It would be great if this could get merged for 1.37.

**Pull request merged**